Miopen dialect opt step8 : Lowering logic for `blockwise_load`/`blockwise_store`/`threadwise_load`/`threadwise_store`.

whchung opened the pull request:
Lower `blockwise_load` to `threadwise_load`, `blockwise_store` to `threadwise_store`.

whchung committed:
Suppress debug messages.

whchung committed:
Lowering logic for blockwise_load/blockwise_store.

whchung committed:
Preliminary implementation of threadwise_load/store lowering logic.

whchung committed:
Fix build errors due to cherry-pick.

whchung committed:
Remove compile-time warnings.

whchung committed:
Rename one variables.

whchung committed:
Remove unused comments.

whchung committed:
Revise comments.

whchung committed:
Reorder logic.

whchung committed:
Comment out logic in threadwise_load / threadwise_store for now.

whchung committed:
Fix comment.

whchung committed:
Refactor logic. Move sourceCoord/destCoord to the logic truly depends on it.

whchung committed:
Extract a lambda to a function.

whchung committed:
Rename a variable.

whchung committed:
Revise comments and logic to improve readability.

whchung committed:
Consolidate some variables used by threadwise_copy and threadwise_copy_v2.

whchung committed:
Remove unused vairables in threadwise_copy.

composedSourceTransform can replace 2 variables.
composedDestTransform can replace 2 other variables.

whchung committed:
Remove unused vairables in threadwise_copy_v2.

composedSourceTransform can replace 2 variables.
composedDestTransform can replace 2 other variables.

whchung committed:
Remove old comments.

whchung committed:
Remove some redundant braces.

whchung committed:
Add comments.

whchung committed:
Extract logic for naive tensor copy to a lambda.

whchung committed:
Refactor logic.

whchung committed:
Refactor logic. Extract part of coordinate computing to a lambda.

whchung committed:
Convert the lambda to a function.

whchung committed:
Refactor logic to collect information into a lambda.

whchung committed:
Refactor information to extract OOB dim logic to a lambda.

whchung committed:
Refactor logic from a lambda to a function.

whchung committed:
Refactor OOB logic from a lambda to a function.

whchung committed:
Add comments.

whchung committed:
Remove some unused comments.

comments remove in this commit were copied C++ logic in the original MIOpen
v4r4 algorithm. They are not relevant now.

some comments which refers to v4r4 and XDLOPS code selection and coordinate
computation logic are still retained for reference. they may be removed in
future commits when those relevant sections are examined / refactored.

whchung committed:
Refactor logic to compute index diff maps to a lambda.

whchung committed:
Refactor logic from a lambda to a function.

whchung committed:
Refactor logic to emit load to a lambda.

whchung committed:
Refactor logic from a lambda to a function.

whchung committed:
Refactor logic to emit store to a lambda.

whchung committed:
Refactor logic from a lambda to a function.

whchung committed:
Merge branch 'miopen-dialect' into miopen-dialect-step9

whchung committed:
Remove compile-time warnings.

whchung committed:
Dogfood extracted functions in threadwise_copy_v2 lowering logic.

whchung committed:
Dogfood extracted functions in threadwise_copy_v2 lowering logic.

whchung committed:
Remove threadwise_load/store lowering logic.

It would be submitted in other PRs.

whchung committed:
Fix compile-time errors and warnings.

whchung committed:
Let blockwise_store / blockwise_load take a vector of 3 elements.

whchung committed:
Revert "Remove threadwise_load/store lowering logic."

This reverts commit 4c02370464f207f009310b89456e168702000d6b.

whchung committed:
Merge branch 'miopen-dialect' into miopen-dialect-step9

whchung committed:
Bug fix: pass oob check dims by reference instead of copy.

whchung committed:
Merge branch 'miopen-dialect' into miopen-dialect-opt-step8

whchung committed:
Remove unused codes.

whchung committed:
Initial take to realize threadwise_load based on infrastructure put in place.

vector load logic hasn't been enabled yet.

whchung committed:
Initial take to realize threadwise_store based on infrastructure put in place.

vector store logic hasn't been enabled yet.

whchung committed:
Initial take to integrate blockwise_load/blockwise_store into gridwise_gemm.

whchung committed:
Initial take to integrate blockwise_load/blockwise_store into gridwise_gemm_v2.

whchung committed:
Fix a bug in gridwise_gemm lowering logic assigning attributes.

whchung committed:
Fix a bug in blockwise_load/store lowering wrt coorindate lengths.

whchung committed:
Revise logic to compute scalar / vector types to be loaded. Start use tuple.

whchung committed:
Change syntax of blockwise/threadwise_load/store to take tuple types.

whchung committed:
Fix one unit test after switching to tuple.

whchung committed:
Fix gridwise_gemm / gridwise_gemm_v2 lowering logic to support tuple.

whchung committed:
Factor out common workaround logic for debugging.

whchung committed:
Change vector.tuple op so it accepts scalar types as well.

whchung committed:
XXX. Experimental commit.

DOES NOT WORK YET!

TBD:

- in threadwise_load, let the return value contain the layout of loaded values.
- pass such layout to threadwise_store to form sliceLengths.
- revise emitLoadLogic, perhaps create an emitLoadInstruction.
- take OOB load into consideration.
- revise emitStoreLogic, perhaps create an emitStoreInstruction.
- take OOB store into consideration.
- revisit the decision to remove convertScalarValue in emitLoadLogic.

- e2e test for scalar cases without padding. f32/f16.
- e2e test for scalar cases with padding. f32/f16.
- e2e test for vector cases without padding. f32/f16.
- e2e test for vector cases with padding. f32/f16.

whchung committed:
Factor out common logic to compute slice lengths in threadwise_copy/v2.

This is a pre-cursor to adopt the same for vectorized threadwise_load/store.

whchung committed:
Refactor the logic and make more verbosed comments.

whchung committed:
WIP commit to get vector load/store logic fully functional.

- set bound attribute for threadwise_load.
- set bound attribute for threadwise_store.
- modify slice length per vector / scalar size.
- modify threadwise_store to extract elements from tuple.
- revise emitStoreLogic
- create an emitStoreInstruction
- take OOB store into consideration.
- revise emitLoadLogic
- create an emitLoadInstruction.
- take OOB load into consideration.
- e2e test for scalar cases without padding. f32/f16.
- e2e test for scalar cases with padding. f32/f16.
- revisit the decision to remove convertScalarValue in emitLoadLogic.

TBD:

- e2e test for vector cases without padding. f32/f16.
- e2e test for vector cases with padding. f32/f16.

Current status:

Need to fix the case where we do vector threadwise_load, but have to do
scalar threadwise_store.

rocm-user@ixt-rack-112:~/llvm-project-mlir/build$
./bin/mlir-miopen-driver -p -pv -rand 1 -rand_side input -c  |
./bin/mlir-rocm-runner
--shared-libs=./lib/librocm-runtime-wrappers.so,./lib/libmlir_runner_utils.so
--entry-point-result=void
Unranked Memref base@ = 0x7ffedde795f0 rank = 1 offset = 0 sizes = [1]
strides = [1] data =
[0]

rocm-user@ixt-rack-112:~/llvm-project-mlir/build$
./bin/mlir-miopen-driver -p -pv -rand 1 -rand_side filter -c  |
./bin/mlir-rocm-runner
 --shared-libs=./lib/librocm-runtime-wrappers.so,./lib/libmlir_runner_utils.so
--entry-point-result=void
Unranked Memref base@ = 0x7fff93419050 rank = 1 offset = 0 sizes = [1]
strides = [1] data =
[1]

whchung committed:
Use references instead of copies.

whchung committed:
Revise computeLoadTypeInfo interfaces.

Use attributes computed in earlier passes.

whchung committed:
Merge branch 'miopen-dialect' into miopen-dialect-opt-step8

whchung committed:
XXX experimental commit.

whchung commented:
After the deep dive with @asroy today, some observations:

- need to break all vector types to scalar types. vectors would only be used transiently in case we wish to actually emit vector load / store instructions. tuples would only carry scalars.
- load/store take place on top-level tensor coordinates, depending on the layout, a vector load might trigger scalar stores, and scalar loads might trigger vector stores.

whchung committed:
XXX experimental commit 2.

decompose loaded vector into scalars.

force scalar stores right now.

TBD:

- change ThreadwiseStoreRewritePattern to be symmetrical with ThreadwiseLoadRewritePattern
  - construct scalars to a vector and pass to emitStoreLogic in case possible (dstDataPerWrite > 1).
- test OOB cases
- test XDLOPS cases

whchung committed:
XXX experimental commit 3.

Access tuples loaded per IVs in threadwise_store.

whchung committed:
XXX experimental commit 4.

access the tuples per IVs in threadwise_load.

whchung committed:
Merge branch 'miopen-dialect' into miopen-dialect-opt-step8

whchung commented:
jenkins: retest this please.

whchung commented:
Adding @asleepzzz as I have to change OOB store logic to switch from `i32` to `index` type to attain symmetry in the implementation.  I expect symmetry inside `emitLoadLogic` and `emitStoreLogic`.

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/Tuning/GridwiseGemmParams.h:
```diff
@@ -363,6 +366,7 @@ class PopulateParamsBase {
       dataPerThreadCopyGemmPos2 = dataPerThreadCopyGemmVectorized;
       derived.srcVectorReadDim = 2;
     }
+    assert(derived.srcVectorReadDim != 0);
```
Does this assertion serve only to make sure you've set `derived.srcVectorReadDim` in each branch of the `if`? If so, would a comment be called for to make sure people know why this is here?

whchung replied:
The value 0,1,2 is pretty bad. I’ll fix them with constant literals

whchung committed:
Fix unit test due to changes in vector.tuple.

Fix unit test due to changed in vector.tuple introduced in commit#
65d52d678945b667271602300d61f529b61bfab8

whchung commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -256,23 +409,49 @@ inline void emitStoreLogic(OpBuilder &b, Location loc, MemRefType destType,
+              b.create<IndexCastOp>(loc, v, b.getIntegerType(32)));
+        b.create<gpu::RawbufStoreOp>(loc, value, dest, oob,
+                                     destLowerIndicesI32);
+      } else {
```
consider user vector write so we could possibly emit `ds_write_b128`.

whchung replied:
defer this to the next PR.

whchung commented:
Current status of the PR:

- ensure all cases which doesn't require OOB checks work fine.
- ensure all cases with OOB checks, without padding dimensions along GemmK/M/N direction work fine.
- ensure all cases with OOB checks, with padding along GemmK dimension work fine.

we should be able to cover all resnet50 configs given the current condition of PR #200 . 

the final 1 case need more work:
- ensure all cases with OOB checks, with padding along GemmM/N direction work fine.

whchung committed:
Merge remote-tracking branch 'origin/miopen-dialect' into miopen-dialect-opt-step8

whchung committed:
Rename variables.

whchung committed:
Add FIXME comment.

whchung committed:
Remove unused variables.

whchung committed:
Rename variable to make it more readable.

whchung committed:
Force scalar store at this moment.

whchung committed:
Rename variables.

whchung committed:
Revise threadwise_store logic to be more symmetic with threadwise_load.

whchung committed:
Add more debug logic.

whchung committed:
Emit OOB store logic properly for scalar types.

whchung committed:
Temporarily set backward data to only use scalar loads/stores.

whchung committed:
Use enum instead of numbers for gemm dimensions.

whchung committed:
Merge branch 'miopen-dialect' into miopen-dialect-opt-step8

whchung committed:
Use enum instead of numbers for gemm dimensions.

whchung committed:
Fix unit tests.

asroy commented on mlir/include/mlir/Dialect/MIOpen/Tuning/GridwiseGemmParams.h:
```diff
@@ -316,15 +320,15 @@ class PopulateParamsBase {
+    bool gemmKVectorizable = false;
     int64_t vectorizableLength = 0;
     if (isGemmA) {
       obtainGemmADimKVectorizable(ctx.opType, ctx.dimIndexVal,
```
I didn't review what's happening inside ```obtainGemmADimKVectorizable```, etc

whchung replied:
it's okay for now. but we will need to review them in subsequent PRs.

asroy commented on mlir/include/mlir/Dialect/MIOpen/Tuning/GridwiseGemmParams.h:
```diff
@@ -341,8 +345,18 @@ class PopulateParamsBase {
+    // backward data has to be reviewed.
+    auto opType = ctx.opType;
+    if (opType == mlir::miopen::ConvOpType::Conv2DBwdDataOpType) {
+      vectorizationSize = 1;
```
vectorizationSize is 1, so it should be safe.

This function (PopulateParamsBase) doesn't have much to be reviewed for now.

obtainGemmADimKVectorizable, etc and vectorizationSize need to be updated in future PR

whchung replied:
we will need to review those logic and file subsequent PRs.

whchung committed:
Temporarily set backward weight to only use scalar loads/stores.

whchung committed:
Fix clang-format.

whchung commented:
jenkins: retest this please.

whchung commented:
Merging the PR as it has been verified on the nightly CI:

http://ml-ci.amd.com:21096/job/MLIR/job/test-step8/5/console

whchung commented:
`rocm-head-ci` has actually passed, see:
http://10.236.106.97:8080/job/mlir/job/PR-200/5/
